### PR TITLE
[RLlib] Bump torch versions and CUDA toolkit

### LINKED ISSUE #22557: [Bug] High-throughput architectures (IMPALA, APPO) regression
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.


### Ray Component

RLlib

### What happened + What you expected to happen

I was doing some experiments, and for some reason I couldn't get APPO to work properly, while I can get PPO to work quite reasonably. This led me to some debugging, and currently to me it seems that APPO might not be working at all, and potentially IMPALA is significantly less efficient than PPO (which might be a separate issue).

Looking at the reward curves (red - PPO, orange - IMPALA, pink - APPO), it seems that:

1. PPO trains the fastest in sample time (x axis - timesteps):
![rewards1](https://user-images.githubusercontent.com/3268882/155228257-6b0dd831-9304-4869-a44f-89c39b6a775b.png)

2. PPO vs IMPALA is almost the same in wall time (x axis - hours):
![rewards2](https://user-images.githubusercontent.com/3268882/155228256-5b47626b-8c0e-4184-ba35-a1595b8a2f8d.png)

3. The sample throughput seems to make sense for PPO and IMPALA, but it's disturbingly low for APPO:
![samples](https://user-images.githubusercontent.com/3268882/155228253-eb38ca5e-5d11-4101-a33e-0cf9778b0069.png)

Also, IMPALA is nowhere near the reported wall time ([here](https://github.com/ray-project/rl-experiments#pong-in-3-minutes)), while I'm using 32 workers. Well, I have 24 cores, and I set the `num_cpus_per_worker` to 0.5, because it doesn't saturate my system anyway, but I can run a benchmark on a 32 core machine if needed:
![cpu](https://user-images.githubusercontent.com/3268882/155229223-db754fde-0a90-4e6a-8527-58356b60c5f8.png)

Also, the GPU utilization seems a bit odd (I have 2 1080ti GPUs, but trained all 3 just on 1 GPU, I don't know why they show up on both graphs):
![gpu](https://user-images.githubusercontent.com/3268882/155229425-11a2dd4e-8da5-4788-881c-e08d801e7533.png)


So, I can partially explain the speed of IMPALA, probably my GPU cannot process the samples fast enough, so the overall training speed is lower than the reported. However, it shouldn't be lower then PPO, but I'm aware that they are different algorithms, so it's not completely apples to apples.

On the other hand, APPO doesn't seem to learn at all, which is odd. Also, to add to that, with APPO the training ends with OOM:
```
Using FIFO scheduling algorithm.                                                                                                                                                                          [52/9617]
Resources requested: 17.0/24 CPUs, 1.0/1 GPUs, 0.0/36.24 GiB heap, 0.0/18.12 GiB objects                                                                                                                           
Result logdir: /home/user/ray_results/pong-appo                                                                                                                                                                   
Number of trials: 1/1 (1 RUNNING)                                                                                                                                                                                  
+-------------------------------------+----------+----------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+                      
| Trial name                          | status   | loc                  |   iter |   total time (s) |      ts |   reward |   episode_reward_max |   episode_reward_min |   episode_len_mean |                      
|-------------------------------------+----------+----------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------|                      
| APPO_PongNoFrameskip-v4_639fd_00000 | RUNNING  | 192.168.1.74:4189099 |   1167 |          13934.2 | 3046000 |   -20.22 |                  -17 |                  -21 |            3747.82 |                      
+-------------------------------------+----------+----------------------+--------+------------------+---------+----------+----------------------+----------------------+--------------------+                      
                                                                                                         
                                                                                                                                                                                                                   
(RolloutWorker pid=4189081) [2022-02-22 21:32:41,252 E 4189081 4189081] plasma_store_provider.cc:122: Failed to put object bbb0ade83413523bb3b6e75bbebcc7d86b1d79350100000001000000 in object store because it is full. Object size is 11313457 bytes.                                                                                                                                                                                
(RolloutWorker pid=4189081) Plasma store status:                                                                                                                                                                   
(RolloutWorker pid=4189081)                                                                                                                                                                                        
(RolloutWorker pid=4189081) (global lru) capacity: 19458202828                                                                                                                                                     
(RolloutWorker pid=4189081) (global lru) used: 0%                                                                                                                                                                  
(RolloutWorker pid=4189081) (global lru) num objects: 0                                                                                                                                                            
(RolloutWorker pid=4189081) (global lru) num evictions: 42313                                                                                                                                                      
(RolloutWorker pid=4189081) (global lru) bytes evicted: 468750773694                                     
(RolloutWorker pid=4189081) ---                                                                          
(RolloutWorker pid=4189081) --- Tip: Use the `ray memory` command to list active objects in the cluster. 
(RolloutWorker pid=4189081) ---                                                                          
(RolloutWorker pid=4189081)                         
(RolloutWorker pid=4189082) [2022-02-22 21:32:41,252 E 4189082 4189082] plasma_store_provider.cc:122: Failed to put object c504bdc6be5ac365cb0b11874cedf833ae26ec1e0100000001000000 in object store because it is full. Object size is 11313457 bytes.                                                                                                                                                                                
(RolloutWorker pid=4189082) Plasma store status:                                                                                                                                                                   
(RolloutWorker pid=4189082)                                                                                                                                                                                        
(RolloutWorker pid=4189082) (global lru) capacity: 19458202828                                                                                                                                                     
(RolloutWorker pid=4189082) (global lru) used: 0%                                                                                                                                                                  
(RolloutWorker pid=4189082) (global lru) num objects: 0                                                                                                                                                            
(RolloutWorker pid=4189082) (global lru) num evictions: 42313                                                                                                                                                      
(RolloutWorker pid=4189082) (global lru) bytes evicted: 468750773694                                                                                                                                               
(RolloutWorker pid=4189082) ---                     
(RolloutWorker pid=4189082) --- Tip: Use the `ray memory` command to list active objects in the cluster. 
(RolloutWorker pid=4189082) ---                                                                          
(RolloutWorker pid=4189082)                                                                              
```

I can reproduce the same behaviour with Tensorflow and Pytorch.


For all these experiment I used:
```
rllib/tuned_examples/impala/pong-impala-fast.yaml
rllib/tuned_examples/ppo/pong-ppo.yaml
rllib/tuned_examples/ppo/pong-appo.yaml
```

### Versions / Dependencies

Ray: 3cb858 (latest master)
Ubuntu: 20.04
Tensorflow: 2.8.0
Pytorch: 1.10.2

### Reproduction script

```
python rllib/train.py -f <exp file>
```
with `exp file`:
```
rllib/tuned_examples/impala/pong-impala-fast.yaml
rllib/tuned_examples/ppo/pong-ppo.yaml
rllib/tuned_examples/ppo/pong-appo.yaml
```

### Anything else

Every time

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

After conducting a row of tests, we have detected a throughput regression in PyTorch, primarily in our APPO release test. 
This has been linked to CUDA toolkit versions != 10.x (meaning CUDA toolkit==11.x and therefore cudnn 8.x). We have tracked this down using the torch profiler to two cudnn operations that use up a lot more time when using CUDA toolkit==11.x and cudnn==8.x.

According to the [release notes](https://docs.nvidia.com/deeplearning/cudnn/release-notes/rel_8.html) of all cuDNN Subversions 8.X.Y, there are a number of performance regressions from cuDNN 7.6.X to be found as “known issues” with cuDNN 8.X.Y..
It is possible that some of these, vaguely formulated, issues apply to our networks and furthermore possible, that we hit undocumented issues with cuDNN.
Torch does not offer deeper interaction with CUDA or cuDNN than what is to see with the used profiler above. We are therefore limited to comparing our observations of version numbers to the release notes. I conclude that on newer hardware, performances between CUDA toolkit 10.X and 11.X / cuDNN 7.X and 8.X seem comparable (with CUDA toolkit 10.x and cuDNN 7.X still leading to the best results in the benchmark). On older hardware, the legacy CUDA toolkit 10.X and cuDNN 7.X outperform by far. Without being able to profile all combinations for the algorithms used by cuDNN, it seems to me that we have to accept that calculations done on our specific networks by cuDNN 7.X are faster than the calculations done by 8.X due to a regression in the cuDNN algorithm selection and/or execution time. As stated above, this effect varies with the used GPU.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ArturNiederfahrenhorst** [comment]
This PR includes both places that a change would be effective in.
1) The docker ml requirements affect all GPU tests and 2) the app config only affects RLLib.

File: rllib/algorithms/r2d2/r2d2_tf_policy.py [hunk 1]
```diff
@@ -207,7 +207,7 @@ def h_function(x, epsilon=1.0):
 
 
 def h_inverse(x, epsilon=1.0):
-    """Inverse if the above h-function, described in the paper [1].
+    """Inverse of the above h-function, described in the paper [1].
```
Comment (gjoliver) [inline rllib/algorithms/r2d2/r2d2_tf_policy.py:210]: this change looks completely unrelated?

File: python/requirements.txt [hunk 2]
```diff
@@ -47,38 +47,45 @@ starlette==0.18.0
 aiorwlock
 gradio; platform_system != "Windows"
 
-# Requirements for running tests
-pyarrow >= 6.0.1, < 7.0.0
-# Used for Dataset tests.
-polars
+## Requirements for running tests
+
+# General test requirements
```
Comment (gjoliver) [inline python/requirements.txt:52]: this comment seems a bit redundant with the one right above.
is this long list of deps all for testing?

**Review by gjoliver — APPROVED** [review]
finally! thanks!

File: rllib/algorithms/r2d2/r2d2_tf_policy.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/r2d2/r2d2_tf_policy.py:210]: This is to trigger library tests, since CI did not kick off library tests if no changes to libraries where made. This might have changed because I reported to Kai that we should trigger tests if requirements files change. I'll have another look.

File: python/requirements.txt (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply python/requirements.txt:52]: Over time, there where to categories of tests requirements added (but they where added right inside these "general" requirements. I've split this up into "General test requirements", "For doc tests" and "For dataset tests". They are all not setup requirements though. That's the logic behind the commet. Happy to change - should I?

File: .buildkite/pipeline.gpu_large.yml [hunk 3]
```diff
@@ -57,5 +57,8 @@
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
     - DOC_TESTING=1 TRAIN_TESTING=1 TUNE_TESTING=1 ./ci/env/install-dependencies.sh
     - pip install -Ur ./python/requirements_ml_docker.txt
+    # (artur) Install torch version that also fits huggingface's usage so that the transformer GPU test passes
```
Comment (ArturNiederfahrenhorst) [inline .buildkite/pipeline.gpu_large.yml:60]: huggingface themselves use 11.1 in their tests afaics. CI fails with a CUDA OOM.

Comment (ArturNiederfahrenhorst) [reply .buildkite/pipeline.gpu_large.yml:60]: @krfricke
